File: ops/dispatch-worker/src/resource-status.js
// Latest recorded usage; never infer a live balance from an old reading.
const stamp = value => {
  const date = new Date(value);
  return Number.isFinite(date.getTime()) ? new Intl.DateTimeFormat('en-GB', {timeZone:'Asia/Kolkata',day:'2-digit',month:'short',year:'numeric',hour:'2-digit',minute:'2-digit',hourCycle:'h23'}).format(date)+' IST' : 'unknown';
};
function midnight(now, zone) {
  const base = Math.floor(now/3600000)*3600000;
  for (let hour=1; hour<=36; hour++) {
    const next=base+hour*3600000;
    if (new Intl.DateTimeFormat('en-GB',{timeZone:zone,hour:'2-digit',hourCycle:'h23'}).format(next)==='00') return next;
  }
  return NaN;
}
export function formatResources(vendors={}, ledger={}, now=Date.now(), apis={}) {
  const lines=['Resources — '+stamp(now)], day=new Date(now).toISOString().slice(0,10), cf=ledger[day];
  if (cf) { const used=(cf.neurons||0)+(cf.text_neurons||0); lines.push(`Cloudflare AI: ${used.toFixed(2)}/10,000 units used; ${Math.max(0,10000-used).toFixed(2)} left (recorded use).`); }
  else lines.push(`Cloudflare AI: no usage record for ${day}; remaining unknown.`);
  lines.push('Cloudflare reset: '+stamp(midnight(now,'UTC'))+'.');
  const gem=vendors.gemini||{};
  lines.push('Gemini data: '+stamp(gem.observed_at)+'.');
  for (const [model, record] of Object.entries(gem.models||{})) lines.push(`  ${model}: ${record.made??'unknown'} requests recorded${record.out_of_quota_at?'; limit reached':''}.`);
  lines.push('Gemini remaining: unknown; no account ceiling reported.','Gemini daily reset: '+stamp(midnight(now,'America/Los_Angeles'))+'.');
  const groq=vendors.groq||{};
  lines.push(`Groq ${groq.model||'model unknown'} — data: ${stamp(groq.observed_at)}.`);
  for (const key of ['requests','tokens']) {
    const reading=groq[key]||{};
    lines.push(`  ${key}: ${reading.remaining??'unknown'}/${reading.limit??'unknown'} left at that reading.`);
    lines.push('  Full refill if unused: '+stamp(reading.reset_seconds==null?NaN:new Date(groq.observed_at).getTime()+reading.reset_seconds*1000)+'.');
  }
  const ig=apis.instagram||{}, cfg=ig.config||{};
  if (typeof ig.quota_usage==='number' && typeof cfg.quota_total==='number') {
    lines.push(`Instagram: ${ig.quota_usage}/${cfg.quota_total} posts used; ${Math.max(0,cfg.quota_total-ig.quota_usage)} left. Data: ${stamp(apis.observed_at)}.`);
    lines.push(`Instagram window: ${cfg.quota_duration??'unknown'} seconds; rolling refill, no fixed daily reset.`);
  } else lines.push('Instagram API allowance: remaining and reset unknown; no reading available.');
  lines.push('Telegram API: daily remaining not reported; retry delays come from its response.','Old readings are not live balances; other use may reduce what remains.');
  return lines.join('\n');
}
export async function resources(env) {
  if (!env.GH_DISPATCH_TOKEN) return formatResources();
  try {
    const values = await Promise.all(['vendor_quotas.json','flux_neurons.json','api_quotas.json'].map(async name => {
      const response=await fetch(`https://api.github.com/repos/mayankav/ig-work/contents/state/${name}?ref=main`,{headers:{Authorization:`Bearer ${env.GH_DISPATCH_TOKEN}`,Accept:'application/vnd.github.raw+json','User-Agent':'suresilly-resources'},signal:AbortSignal.timeout(4000)});
      if (!response.ok) return {};
      return response.json();
    }));
    return formatResources(values[0],values[1],Date.now(),values[2]);
  } catch { return formatResources()+'\nLatest saved data could not be read.'; }
}

File: scripts/resource_status.py
"""Read-only resource report. Counts are not invented remaining allowances."""
import json
import os
from datetime import datetime, timedelta, timezone
from pathlib import Path
from zoneinfo import ZoneInfo
ROOT = Path(__file__).resolve().parents[1]
IST = ZoneInfo('Asia/Kolkata')


def read(path):
    try: return json.loads(path.read_text())
    except (OSError, ValueError): return {}


def stamp(value):
    try:
        date = value if isinstance(value, datetime) else datetime.fromisoformat(value.replace('Z','+00:00'))
        return date.astimezone(IST).strftime('%d %b %Y %H:%M IST')
    except (TypeError, ValueError, AttributeError): return 'unknown'


def refresh_api():
    """Read Instagram's account limit; missing access is not a zero balance."""
    key, user = os.environ.get('IG_ACCESS_TOKEN'), os.environ.get('IG_USER_ID')
    if not key or not user: return {}
    import requests
    from instagram_api import graph_base
    result = {'observed_at': datetime.now(timezone.utc).isoformat(), 'instagram': {}}
    try:
        response = requests.get(f'{graph_base(key)}/{user}/content_publishing_limit',
            params={'fields':'quota_usage,config','access_token':key}, timeout=5)
        if response.status_code == 200:
            data = response.json().get('data', [])
            if data: result['instagram'] = {k:data[0].get(k) for k in ('quota_usage','config')}
        else: result['status'] = 'API reading unavailable'
    except Exception: result['status'] = 'API reading unavailable'
    return result


def save_api():
    data = refresh_api()
    if data: (ROOT/'state/api_quotas.json').write_text(json.dumps(data,indent=2)+'\n')
    return data


def report(vendors=None, ledger=None, now=None, apis=None):
    now = now or datetime.now(timezone.utc)
    vendors = read(ROOT/'state/vendor_quotas.json') if vendors is None else vendors
    ledger = read(ROOT/'state/flux_neurons.json') if ledger is None else ledger
    day = now.astimezone(timezone.utc).date().isoformat()
    cf = ledger.get(day)
    lines = ['Resources — '+stamp(now)]
    if isinstance(cf,dict):
        spent = cf.get('neurons',0)+cf.get('text_neurons',0)
        lines.append(f'Cloudflare AI: {spent:,.2f}/10,000 units used; {max(0,10000-spent):,.2f} left (recorded use).')
    else: lines.append('Cloudflare AI: no usage record for '+day+'; remaining unknown.')
    lines.append('Cloudflare reset: '+stamp((now+timedelta(days=1)).replace(hour=0,minute=0,second=0,microsecond=0))+'.')
    gem = vendors.get('gemini',{})
    lines.append('Gemini data: '+stamp(gem.get('observed_at'))+'.')
    for model,record in gem.get('models',{}).items():
        lines.append(f"  {model}: {record.get('made','unknown')} requests recorded"+('; limit reached' if record.get('out_of_quota_at') else '')+'.')
    lines.append('Gemini remaining: unknown; no account ceiling reported.')
    pacific=now.astimezone(ZoneInfo('America/Los_Angeles'))
    lines.append('Gemini daily reset: '+stamp((pacific+timedelta(days=1)).replace(hour=0,minute=0,second=0,microsecond=0))+'.')
    groq=vendors.get('groq',{})
    lines.append('Groq '+groq.get('model','model unknown')+' — data: '+stamp(groq.get('observed_at'))+'.')
    for key in ('requests','tokens'):
        reading=groq.get(key,{})
        lines.append(f"  {key}: {reading.get('remaining','unknown')}/{reading.get('limit','unknown')} left at that reading.")
        try:
            refill=datetime.fromisoformat(groq['observed_at'].replace('Z','+00:00'))+timedelta(seconds=reading['reset_seconds'])
            lines.append('  Full refill if unused: '+stamp(refill)+'.')
        except (KeyError,ValueError,TypeError): lines.append('  Refill time: unknown.')
    apis = read(ROOT/'state/api_quotas.json') if apis is None else apis
    ig = apis.get('instagram', {}); config = ig.get('config') or {}
    used, total = ig.get('quota_usage'), config.get('quota_total')
    if isinstance(used, (int,float)) and isinstance(total, (int,float)):
        lines.append(f'Instagram: {used}/{total} posts used; {max(0,total-used)} left. Data: '+stamp(apis.get('observed_at'))+'.')
        lines.append(f"Instagram window: {config.get('quota_duration','unknown')} seconds; rolling refill, no fixed daily reset.")
    else: lines.append('Instagram API allowance: remaining and reset unknown; no reading available.')
    lines.append('Telegram API: daily remaining not reported; any retry delay is handled from its response.')
    lines.append('Old readings are not live balances; other use may reduce what remains.')
    return '\n'.join(lines)

if __name__=='__main__': print(report())
